game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java

```diff
@@ -69,6 +69,256 @@
     PARTIAL_AMPHIB
   }
 
+  /**
+   * An action representing attacking aa guns firing during a battle.
+   *
+   * <p>NOTE: This type exists solely for tests to interrogate the execution stack looking for an
+   * action of this type.
+   */
+  public static abstract class FireOffensiveAaGuns implements IExecutable {
```

Comment:
Marker interfaces/classes are usually an anti-pattern. Are each of these steps associated with a specific battle step string?

It seems like this is perhaps more fitting to an enum. EG:
```
@AllArgsConstructor
enum BattleStepPhase {
  FIRE_OFFENSIVE_AA_GUNS("Fire Offensive AA"),
  :
  :
  ;

  @Getter
  private final String label;
}
```

At that point the battle step list can become a list of `BattleStepPhase` and then just convert it to a list of strings at time of display, eg:
```
  battleStepPhases.stream()
     .map(BattleStepPhase::getLabel)
     .collect(Collectors.toList());
```

With a list of `BattleStepPhase`, you can query that the list has the expected enum element.

To tie it in with the executable, a helper object probably would be needed and you would have a list of those instead. EG:

```
@Value
@Builder
class BattleStep {
   BattleStepPhase battleStepPhase;
   Runnable executable;
}
```

Then you'd have `List<BattleStep>`, and you could query which battle phase it is, get the label from that, and have a reference to the actions to be run. WDYT?

Reply:
This is basically what I have in mind.  But you keep saying do small PRs :)

It isn't quite an enum.  The order of steps is dynamic depending on the units and depending on the state of the battle.  And a "step" can have zero or more "step strings".  Also a "step string" might be shown but the "step" actually won't run because of changes to the units during the course of the battle.

This PR is just go get started with unittests in the code base and to work out any issues around those unittests.  I don't want to make any structural changes at this point for fear of breaking something without realizing it.

Reply:
Understood, there is a trade-off.

These classes being present only for tests is not ideal:
```
  public abstract static class AttackerRetreatPlanes implements IExecutable {
    private static final long serialVersionUID = 6828071620685651390L;
  }
```

I can understand that being an intermediary step for lack of better solutions without going into deeper changes. I wonder though if the unit tests are going to be re-written just as well if you break up the code. If when re-writing the prod code you also have to re-write the tests, then it's still a re-write nonetheless. Unit tests written after-the-fact can suffer from verifying (or testing) that a bug exists! IE: the code could be incorrect, and there can be a test that asserts the code functions as-is, bugs included. 

If you break up the battle steps to their own classes, they will lose that abstract class and then the test will break because it's asserting for the wrong abstract class.